README.md
# Express Route Keeper

__Important!__ This project is in development, README.md is draft API.

This project is a route middleware for acls and parameter checker.

## Installation

```
npm i express-route-keeper
```

## Usage

```javascript
import express from 'express';
import bodyParser from 'body-parser';
import RouteKeeper from 'express-route-keeper';

const app = express();
const keeper = RouteKeeper();

const READ_PROJECT = Symbol('READ_PROJECT');
const CREATE_PROJECT = Symbol('CREATE_PROJECT');
const PROJECT_MASTER = Symbol('PROJECT_MASTER');

// middleware for request body parse
app.use(bodyParser.urlencoded({
  extended: true,
}));
app.use(bodyParser.json());

// acl role injection
app.use((req, res, next) => {
  req.aclActions = [
    READ_PROJECT,
    CREATE_PROJECT,
    PROJECT_MASTER
  ];

  next();
});

// public route
app.get('/news', routeHandler);

// acl route
app.get('/projects', keeper(READ_PROJECT), routeHandler);
app.get('/projects', keeper([
  READ_PROJECT,
  PROJECT_MASTER
]), routeHandler); // AND match

// parameter checker
app.post('/login', keeper({
  username: String,
  password: String,
  memorize: {
    type: Boolean,
    defaultValue: false,
    required: false,
  },
}), routeHandler);

// mixed mode
app.post('/projects', keeper({
  acl: [
    CREATE_PROJECT,
    PROJECT_MASTER
  ],
  name: String,
}), routeHandler);

app.listen(PORT);
```

### ACL OR match

```javascript
app.post('/projects', ({
  acl: {
    actions: [
      CREATE_PROJECT,
      PROJECT_MASTER
    ],
    $or: true,
  },
  name: String,
}), routeHandler);
```

### Working with other middleware

```javascript
// General Usage
// Check ACL before uploaded file handler
app.post('/projects', keeper(CREATE_PROJECT), multer.single('cover'), routeHandler);

// Body field checker with multipart
app.post('/projects', multer.single('cover'), keeper({
  name: String,
  managerId: Number,
  acl: [
    CREATE_PROJECT
  ],
}), routeHandler);
```

### Set custom exception

```javascript
import RouteKeeper, {
  PARAMS_ERROR,
  ACL_ERROR
} from 'express-route-keeper';

const keeper = RouteKeeper((err, req, res) => {
  switch (err) {
    case PARAMS_ERROR:
      res.status(401);
      res.json({
        message: 'Invalid Parameter',
      });
      break;

    case ACL_ERROR:
      res.status(401);
      res.json({
        message: 'Authentication Failed',
      });
      break;

    default:
      console.error(err);

      res.status(500);
      res.json({
        message: 'Server Error'
      });
      break;
  }
});
```

### Action Keys

You can use String or ES6 Symbol (Recommend) for action keys.

### Querystring or Body

If request method is __GET__, check the querystring in ```req.query``` or check parameters in ```req.body```.


CHANGELOG.md
# Changelog

## v0.1.4 - 2016-10-26

### Changed

* Change error message from stderr to debug module

## v0.1.3 - 2016-09-20

### Fixed

* Fixed 0 or false value fallback bug

## v0.1.2 - 2016-09-20

### Fixed

* Add json type Boolean parsed true and false detected


lib/routeKeeper.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PARAMS_ERROR = exports.ACL_ERROR = undefined;

exports.default = function (options) {
  if (options) {
    if (_lodash2.default.isFunction(options)) {
      (0, _errorHandler2.setHandler)(options);
    } else if (_lodash2.default.isObject(options)) {
      (0, _errorHandler2.setHandler)(options.onError || _errorHandler);
    }
  }

  return _middlewareMaker2.default;
};

require('core-js/shim');

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _errorHandler2 = require('./errorHandler.js');

var _middlewareMaker = require('./middlewareMaker.js');

var _middlewareMaker2 = _interopRequireDefault(_middlewareMaker);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ACL_ERROR = exports.ACL_ERROR = _errorHandler2.ACL_ERROR;
var PARAMS_ERROR = exports.PARAMS_ERROR = _errorHandler2.PARAMS_ERROR;

;

lib/validParams.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

exports.default = function (rules) {
  var body = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

  _lodash2.default.forEach(rules, function (raw, key) {
    var rule = (typeof raw === 'undefined' ? 'undefined' : _typeof(raw)) !== 'object' ? { type: raw } : raw;

    body[key] = ~[undefined, null].indexOf(body[key]) ? rule.defaultValue : body[key];
    var val = body[key];

    // Skip undefined
    if (rule.required === false && val === undefined) {
      return true;
    }

    switch (rule.type) {
      case Object:
        if (!_lodash2.default.isObject(val)) {
          throw new Error('param `' + key + '` should be Object');
        }
        break;

      case String:
        if (!_lodash2.default.isString(val)) {
          throw new Error('param `' + key + '` should be String');
        }
        break;

      case Array:
        if (!_lodash2.default.isArray(val)) {
          throw new Error('param `' + key + '` should be Array');
        }
        break;

      case Number:
        if (!isNaN(val)) {
          body[key] = val = parseFloat(val);
        }

        if (!_lodash2.default.isNumber(val)) {
          throw new Error('param `' + key + '` should be Number');
        }
        break;

      case Boolean:
        if (~['true', 'false', true, false].indexOf(val)) {
          if (val === 'true' || val === true) {
            body[key] = val = true;
          } else {
            body[key] = val = false;
          }
        }
        if (!_lodash2.default.isBoolean(val)) {
          throw new Error('param `' + key + '` should be Boolean');
        }
        break;
    }
  });
};

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

;

lib/middlewareMaker.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

exports.default = function (options) {
  var errorRoute = (0, _errorHandler.getHandler)();
  var _options = options;

  return function (req, res, next) {
    var aclMode = ACL_AND_MODE;
    var acl = _lodash2.default.isArray(options) ? options : _lodash2.default.isSymbol(options) || _lodash2.default.isString(options) ? [options] : [];
    var params = _lodash2.default.isObject(options) ? _lodash2.default.cloneDeep(options) : null;

    if (params && params.acl) {
      // ACL Params
      var aclRaw = params.acl;
      if (_lodash2.default.isArray(aclRaw)) {
        acl = aclRaw;
      } else if (_lodash2.default.isSymbol(aclRaw) || _lodash2.default.isString(aclRaw)) {
        acl.push(aclRaw);
      } else if (_lodash2.default.isObject(aclRaw) && aclRaw.actions) {
        acl = aclRaw.actions;
        aclMode = aclRaw.$or ? ACL_OR_MODE : ACL_AND_MODE;
      } else {
        throw new Error('acl should be String, Symbol, Object or Array.');
      }

      delete params.acl;
    }

    req.aclActions = req.aclActions || [];

    // ACL Validate
    var notValidAction = null;
    switch (aclMode) {
      case ACL_OR_MODE:
        if (!_lodash2.default.intersection(acl, req.aclActions).length) {
          debugACL('ACL: need at least one action in: ' + acl.map(function (action) {
            return action.toString();
          }).join(', '));
          return errorRoute(_errorHandler.ACL_ERROR, req, res);
        }
        break;

      case ACL_AND_MODE:
        if (acl.map(function (action) {
          return ~req.aclActions.indexOf(action);
        }).some(function (r, index) {
          notValidAction = acl[index].toString();
          return !r;
        })) {
          debugACL('ACL: need action: ' + notValidAction);
          return errorRoute(_errorHandler.ACL_ERROR, req, res);
        }
        break;
    }

    // Parameters Validate
    if (params) {
      try {
        (0, _validParams2.default)(params, req.method === 'GET' ? req.query : req.body);
      } catch (e) {
        debugParams(e);
        return errorRoute(_errorHandler.PARAMS_ERROR, req, res);
      }
    }

    next();
  };
};

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _debug = require('debug');

var _debug2 = _interopRequireDefault(_debug);

var _errorHandler = require('./errorHandler.js');

var _validParams = require('./validParams.js');

var _validParams2 = _interopRequireDefault(_validParams);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ACL_AND_MODE = Symbol('ACL_AND_MODE');
var ACL_OR_MODE = Symbol('ACL_OR_MODE');

var debugACL = (0, _debug2.default)('RouteKeeper:ACL');
var debugParams = (0, _debug2.default)('RouteKeeper:Parameters');

;

lib/errorHandler.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PARAMS_ERROR = exports.ACL_ERROR = undefined;
exports.setHandler = setHandler;
exports.getHandler = getHandler;

var _debug = require('debug');

var _debug2 = _interopRequireDefault(_debug);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ACL_ERROR = exports.ACL_ERROR = new Error('RouteKeeperACLError');
ACL_ERROR.name = 'RouteKeeperACLError';

var PARAMS_ERROR = exports.PARAMS_ERROR = new Error('RouteKeeperParameterError');
PARAMS_ERROR.name = 'RouteKeeperParameterError';

var debugErr = (0, _debug2.default)('RouteKeeper:Error');

function defaultErrorHandler(err, req, res) {
  switch (err) {
    case PARAMS_ERROR:
      res.status(401);
      res.json({
        message: 'Invalid Parameter'
      });
      break;

    case ACL_ERROR:
      res.status(401);
      res.json({
        message: 'Authentication Failed'
      });
      break;

    default:
      debugErr(err);

      res.status(500);
      res.json({
        message: 'Server Error'
      });
      break;
  }
};

var _handler = defaultErrorHandler;

function setHandler(handler) {
  _handler = handler;
};

function getHandler() {
  return _handler;
};